subprojects/test-runner/client/src/main/kotlin/com/avito/runner/scheduler/metrics/TestMetricsAggregatorImpl.kt

```diff
@@ -1,66 +1,70 @@
+        .map { it.finishTime }
         .maxOrNull()
+        .toResult { "Cannot calculate last ended test time" }
 
-    private val queueTimes: List<Long> = testTimestamps
-        .mapNotNull { it.onDevice }
-        .map { it - testSuiteStartedTime }
+    private val queueTimes: List<Long> = testTimestamps.filterIsInstance<TestTimestamps.Finished>()
+        .map { it.onDevice - testSuiteStartedTime }
 
-    private val installationTimes: List<Long> = testTimestamps
-        .mapNotNull { it.installationTime }
+    private val installationTimes: List<Long> = testTimestamps.filterIsInstance<TestTimestamps.Finished>()
+        .map { it.installationTime }
 
-    override fun initialDelay(): Long? = firstTestStarted?.let { it - testSuiteStartedTime }
+    override fun initialDelay(): Result<Long> = firstTestStarted.map { it - testSuiteStartedTime }
 
-    override fun endDelay(): Long? = lastTestEnded?.let { testSuiteEndedTime - it }
+    override fun endDelay(): Result<Long> = lastTestEnded.map { testSuiteEndedTime - it }
 
-    override fun medianQueueTime(): Long? = queueTimes.aggregateOrNull { it.median() }
+    override fun medianQueueTime(): Result<Long> = queueTimes.aggregate(
+        { it.median() },
+        { "Cannot calculate median queue time" }
+    )
 
-    override fun medianInstallationTime(): Long? = installationTimes.aggregateOrNull { it.median() }
+    override fun medianInstallationTime(): Result<Long> = installationTimes.aggregate(
+        { it.median() },
+        { "Cannot calculate median installation time" }
+    )
 
-    override fun suiteTime(): Long? = if (lastTestEnded != null && firstTestStarted != null) {
-        lastTestEnded - firstTestStarted
-    } else {
-        null
-    }
+    override fun suiteTime(): Result<Long> = lastTestEnded.combine(firstTestStarted) { last, first -> last - first }
 
     override fun totalTime() = testSuiteEndedTime - testSuiteStartedTime
 
-    override fun medianDeviceUtilization(): Long? =
-        deviceTimestamps.values
-            .mapNotNull { it.utilizationPercent }
-            .aggregateOrNull { it.median() }
-
-    /**
-     * return null if no data
-     */
-    private fun List<Number>.aggregateOrNull(aggregateFunc: (List<Number>) -> Number): Long? {
-        return if (isNotEmpty()) {
-            val result = aggregateFunc.invoke(this).toLong()
-            if (result > 0) {
-                result
-            } else {
-                null
-            }
-        } else {
-            null
+    override fun medianDeviceUtilization(): Result<Long> =
+        deviceTimestamps.values.filterIsInstance<DeviceTimestamps.Finished>()
+            .map { it.utilizationPercent }
+            .aggregate({ it.median() }) { "Cannot calculate median device utilization" }
+
+    private inline fun Long?.toResult(lazyMessage: () -> String): Result<Long> = when (this) {
+        null -> Result.Failure(IllegalArgumentException(lazyMessage()))
```

Comment:
Maybe `IllegalStateException` will be more precise. It may happen because of missing listener invocation and so on.

Reply:
Hm, it makes sense to me.